File: source/RansomNote.cpp
// https://leetcode.com/problems/ransom-note
#include <string>

#include <unordered_map>

class Solution
{
public:
    bool canConstruct(std::string ransomNote, std::string magazine)
    {
        std::unordered_map<char, int> letters;
        for (const char& c : magazine) ++letters[c];
        for (const char& c : ransomNote)
        {
            --letters[c];
            if (letters[c] < 0) return false;
        }
        return true;
    }
};

File: CSharpPractice/source/RansomNote.cs
// https://leetcode.com/problems/ransom-note
using System.Collections.Generic;

namespace LeetCodePractice.RansomNote
{
    public class Solution
    {
        public bool CanConstruct(string ransomNote, string magazine)
        {
            Dictionary<char, int> letters = new Dictionary<char, int>();
            foreach (char c in magazine)
            {
                if (letters.ContainsKey(c))
                {
                    ++letters[c];
                }
                else
                {
                    letters[c] = 1;
                }
            }
            foreach (char c in ransomNote)
            {
                if (letters.ContainsKey(c))
                {
                    --letters[c];
                    if (letters[c] < 0) return false;
                }
                else
                {
                    return false; // character didnt appear in first string
                }
            }
            return true;
        }
    }
}